ddtrace/appsec/_exploit_prevention/stack_traces.py

```diff
@@ -0,0 +1,74 @@
+import inspect
+from itertools import chain
+from secrets import choice
+from typing import Any
+from typing import Dict
+from typing import Iterable
+from typing import List
+from typing import Optional
+
+from ddtrace._trace.span import Span
+from ddtrace.appsec._constants import EXPLOIT_PREVENTION
+from ddtrace.settings.asm import config as asm_config
+import ddtrace.tracer
+
+
+# remove_top default value of 8
+# report_stack/_waf_action/waf_callable/call_waf_callback
+# wrapped_open/rasp/_wrap_call/patch_func
+
+
+def report_stack(message: str, span: Optional[Span] = None, crop_stack: Optional[str] = None) -> Optional[str]:
+    """
+    Report a stack trace to the current span.
+    This is used to report stack traces for exploit prevention.
+    Return the stack id for the reported stack trace to link it in triggers.
+    """
+    if not asm_config._ep_enabled or not asm_config._ep_stack_trace_enabled:
+        return None
+    if span is None:
+        span = ddtrace.tracer.current_span()
+    if span is None:
+        return None
+    root_span = span._local_root or span
+
+    appsec_traces = root_span.get_struct_tag(EXPLOIT_PREVENTION.STACK_TRACES) or {}
+    exploit: List[Any] = appsec_traces.get("exploit", [])
+
+    # Do not report more than the maximum number of stack traces
+    if asm_config._ep_max_stack_traces and len(exploit) >= asm_config._ep_max_stack_traces:
+        return None
+
+    stack = inspect.stack()
```

Comment:
Just wondering the cost of getting the stacktrace, it may be worth moving this to C at some point, as we did for IAST

Reply:
Exploit prevention should only trigger when a threat meets a vuln. Having a very tiny overhead at that point should be the least of the problem of the customer.